denonavr/api.py

```diff
@@ -316,134 +318,210 @@ def is_default_async_client(self) -> bool:
+        """Establish a connection to the receiver."""
+        loop = asyncio.get_event_loop()
+        _LOGGER.debug("%s: establishing telnet connection", self.host)
+        try:
+            async with async_timeout.timeout(self.timeout):
+                transport_protocol = await loop.create_connection(
+                    lambda: DenonAVRTelnetProtocol(
+                        on_connection_lost=self._disconnected,
+                        on_message=self._process_event
+                    ),
+                    self.host,
+                    23
+                )
+        except asyncio.TimeoutError as err:
+            _LOGGER.debug("%s: Timeout exception on telnet connect", self.host)
+            raise AvrTimoutError(
+                "TimeoutException: {}".format(err),
+                "telnet connect") from err
+        except ConnectionRefusedError as err:
+            _LOGGER.debug(
+                "%s: Connection refused on telnet connect",
+                self.host,
+                exc_info=True
+            )
+            raise AvrNetworkError(
+                "ConnectionRefusedError: {}".format(err),
+                "telnet connect") from err
+        except (OSError, IOError) as err:
+            _LOGGER.debug(
+                "%s: Connection failed on telnet reconnect",
+                self.host,
+                exc_info=True
+            )
+            raise AvrNetworkError(
+                "OSError: {}".format(err), "telnet connect") from err
+        _LOGGER.debug("%s: telnet connection complete", self.host)
+        self._protocol = cast(DenonAVRTelnetProtocol, transport_protocol[1])
+        self._connection_enabled = True
+        self._last_message_time = time.monotonic()
+        self._schedule_monitor()
+
+    def _schedule_monitor(self) -> None:
+        """Start the monitor task."""
+        loop = asyncio.get_event_loop()
+        self._monitor_handle = loop.call_later(
+            _MONITOR_INTERVAL,
+            self._monitor
+        )
+
+    def _stop_monitor(self) -> None:
+        """Stop the monitor task."""
+        if self._monitor_handle is not None:
+            self._monitor_handle.cancel()
+            self._monitor_handle = None
+
+    def _monitor(self) -> None:
+        """Monitor the connection."""
+        time_since_response = time.monotonic() - self._last_message_time
+        if time_since_response > _MONITOR_INTERVAL * 2:
+            _LOGGER.info(
+                "%s: Keep alive failed, disconnecting and reconnecting",
+                self.host
+            )
+            if self._protocol:
+                self._protocol.close()
+            self._disconnected()
+            return
+
+        if time_since_response > _MONITOR_INTERVAL and self._protocol:
+            # Keep the connection alive
+            _LOGGER.debug("%s: Sending keep alive", self.host)
+
+            # Use a command that won't trigger any callbacks
+            self._protocol.write("SLP?\r")
```

Comment:
I don't see any advantage in this. Processing the `power` callback is not expensive. On the contrary, this command might not be supported by all receivers.
```suggestion
            self._protocol.write("PW?\r")
```

Reply:
This is an easy change if you feel strongly about it. My rationale was this - agreed, triggering a callback is inexpensive. However, what this means is the consumer will receive a message indicating a state change every 30 seconds. So in the case of HA, it'll mean triggering a state change and all that good stuff. Just seemed like overhead. The SLP command was added to the Denon protocol at least as far back as 2009. https://assets.denon.com/DocumentMaster/UK/436_AVR4810CI_4810_PROTOCOL_V660a.pdf but it did NOT exist back in 2006 http://assets.eu.denon.com/DocumentMaster/FR/AVR4306_PROTOCOL_V4.6a.pdf  

Denon doesn't seem to provide a list of which models support the timer, without me going through model by model (see https://support-uk.denon.com/app/answers/detail/a_id/1173/~/which-denon-devices-provide-timer-functions%3F)

So I changed this back to PW? per your request in https://github.com/dcmeglio/denonavr/commit/5c5aa1cdcf95be912318d593085b9c2abecab4e3